### PR TITLE
Implement JuMP.value for nonlinear expressions

### LINKED ISSUE #1562: getvalue for nonlinear expressions
https://github.com/JuliaOpt/JuMP.jl/blob/8c36bf2a850af6392a8bba3b8241d3577b69e38a/src/nlp.jl#L1002-L1004

### PR DESCRIPTION
**Pull request by mlubin**
Closes #1562

### CODE REVIEW

File: src/nlp.jl [hunk 1]
```diff
@@ -1025,47 +1025,68 @@ function MOI.constraint_expr(d::NLPEvaluator,i::Integer)
...
+
+    moi_index_to_consecutive_index = Dict{MOI.VariableIndex, Int}()
+    variable_indices = MOI.get(model, MOI.ListOfVariableIndices())
+    variable_values = Array{Float64}(undef, length(variable_indices))
+    for (consecutive_index, moi_index) in enumerate(variable_indices)
+        moi_index_to_consecutive_index[moi_index] = consecutive_index
+        variable_values[consecutive_index] = map(VariableRef(model, moi_index))
+    end
+
+    subexpressions = Array{Vector{NodeData}}(undef,0)
```
Comment (odow) [inline src/nlp.jl]: Nit: space after `,`

File: src/nlp.jl [hunk 2]
```diff
@@ -1025,47 +1025,68 @@ function MOI.constraint_expr(d::NLPEvaluator,i::Integer)
...
+    model = ex.m
+
+    nlp_data::NLPData = model.nlp_data
+
+    moi_index_to_consecutive_index = Dict{MOI.VariableIndex, Int}()
+    variable_indices = MOI.get(model, MOI.ListOfVariableIndices())
+    variable_values = Array{Float64}(undef, length(variable_indices))
+    for (consecutive_index, moi_index) in enumerate(variable_indices)
+        moi_index_to_consecutive_index[moi_index] = consecutive_index
+        variable_values[consecutive_index] = map(VariableRef(model, moi_index))
```
Comment (odow) [inline src/nlp.jl]: I'm not a fan of naming the argument `map` since it makes this pretty ambiguous.

File: test/nlp.jl [hunk 3]
```diff
@@ -568,4 +568,16 @@
             @test_throws expected_exception MOI.eval_objective(d, [1.0, 1.0])
         end
     end
+
+    @testset "JuMP.value on NonlinearExpressions" begin
+        model = Model()
+        @variable(model, x)
+        @NLexpression(model, ex1, sin(x))
+        @NLexpression(model, ex2, ex1 + x^2)
+        @NLexpression(model, ex3, 2ex1 + ex2/2)
```
Comment (odow) [inline test/nlp.jl]: Nit: spaces around binary operators.

File: src/aff_expr.jl [hunk 4]
```diff
@@ -98,16 +98,16 @@ end
 Base.sizehint!(a::GenericAffExpr, n::Int) = sizehint!(a.terms, n)
 
 """
-    value(a::GenericAffExpr, map::Function)
+    value(a::GenericAffExpr, var_value::Function)
```
Comment (blegat) [inline src/aff_expr.jl]: `a` should be renamed `ex`

File: src/aff_expr.jl [hunk 5]
```diff
@@ -98,16 +98,16 @@ end
...
 
-Evaluate `a` given the value `map(v)` for each variable `v`.
+Evaluate `ex` using `var_value(v)` as the value for each variable `v`.
 """
-function value(a::GenericAffExpr{T, V}, map::Function) where {T, V}
-    S = Base.promote_op(map, V)
+function value(a::GenericAffExpr{T, V}, var_value::Function) where {T, V}
```
Comment (blegat) [inline src/aff_expr.jl]: `a` should be renamed `ex`

File: src/nlp.jl [hunk 6]
```diff
@@ -1025,47 +1025,69 @@ function MOI.constraint_expr(d::NLPEvaluator,i::Integer)
...
+    variable_indices = MOI.get(model, MOI.ListOfVariableIndices())
+    variable_values = Array{Float64}(undef, length(variable_indices))
+    for (consecutive_index, moi_index) in enumerate(variable_indices)
+        moi_index_to_consecutive_index[moi_index] = consecutive_index
+        jump_var = VariableRef(model, moi_index)
+        variable_values[consecutive_index] = var_value(jump_var)
+    end
+
+    subexpressions = Array{Vector{NodeData}}(undef, 0)
+    for nl_expr in nlp_data.nlexpr
```
Comment (blegat) [inline src/nlp.jl]: Why not just `Vector{NodeData}[replace_moi_variables(nl_expr.nd, moi_index_to_consecutive_index) for nl_expr in nlp_data.nlexpr]` ?

File: src/nlp.jl [hunk 7]
```diff
@@ -1025,47 +1025,69 @@ function MOI.constraint_expr(d::NLPEvaluator,i::Integer)
...
+    main_expressions = Vector{NodeData}[ex_nd]
+    subexpression_order, _ = order_subexpressions(main_expressions,
+                                                  subexpressions)
+
+    max_len = length(ex_nd)
+    for k in subexpression_order
+        max_len = max(max_len, length(subexpressions[k]))
+    end
+
+    subexpr_values = Array{Float64}(undef, length(subexpressions))
```
Comment (blegat) [inline src/nlp.jl]: Using `Vector` is more readable than `Array` since we know that we are building a vector

File: src/nlp.jl (hunk 6 above)
Comment (mlubin) [reply src/nlp.jl]: It seemed awkward with the line breaks.

**Pull request merged**